tree/dataframe/test/dataframe_datasetspec.cxx

```diff
@@ -194,36 +202,152 @@ TEST(RDFDatasetSpec, MultipleFiles)
    EXPECT_EQ(dfRDS3, dfRange);
+
+   const auto dfRDS4 =
+      RDataFrame(RDatasetSpec({"treeA"s, "treeA"s}, {"file0.root"s, "file1.root"s}, {1, 2})).Display()->AsString();
    EXPECT_EQ(dfRDS4, dfRange);
+
+   const auto dfRDS5 =
+      RDataFrame(RDatasetSpec({"treeA"s, "treeB"s}, {"file1.root"s, "file2.root"s}, {1, 2})).Display()->AsString();
    EXPECT_EQ(dfRDS5, dfRange);
 
+   // specify irregular range [6, 7) (similar to above)
+   auto dfRDS8 = RDataFrame(RDatasetSpec("treeA", {"file0.root"s, "file1.root"s}, {6, 7})).Display()->AsString();
+   // std::string dfRDS8AsString;
+   // ROOT_EXPECT_ERROR(dfRDS8AsString = dfRDS8->AsString(), "TTreeReader::SetEntriesRange()",
+   //                   "first entry out of range 0..6");
+   EXPECT_EQ(dfRDS8, dfEmpty); // ?
+
+   // specify irrelgular ranges (similar to above): [2, 2), [7, 7), [2, 6), [2, 0), [9, 7), [9, 2)
+   const auto dfRDS9 = RDataFrame(RDatasetSpec("treeA", {"file0.root"s, "file1.root"s}, {2, 2})).Display()->AsString();
+   EXPECT_EQ(dfRDS9, dfEmpty);
+
+   const auto dfRDS10 = RDataFrame(RDatasetSpec("treeA", {"file0.root"s, "file1.root"s}, {7, 7})).Display()->AsString();
+   EXPECT_EQ(dfRDS10, dfEmpty);
+
+   const auto dfRDS11 = RDataFrame(RDatasetSpec("treeA", {"file0.root"s, "file1.root"s}, {2, 7})).Display()->AsString();
+   EXPECT_EQ(dfRDS11, dfRange1);
+
+   const auto dfRDSX = RDataFrame(RDatasetSpec("treeA", {"file0.root"s, "file1.root"s}, {2})).Display()->AsString();
+   EXPECT_EQ(dfRDSX, dfRange2);
+
+   auto dfRDSY = RDataFrame(RDatasetSpec("treeA", {"file0.root"s, "file1.root"s}, {6})).Display()->AsString();
+   // std::string dfRDSYAsString;
+   // ROOT_EXPECT_ERROR(dfRDSYAsString = dfRDSY->AsString(), "TTreeReader::SetEntriesRange()",
+   //                   "first entry out of range 0..5");
+   EXPECT_EQ(dfRDSY, dfSimple);
+
+   auto dfRDSZ = RDataFrame(RDatasetSpec("treeA", {"file0.root"s, "file1.root"s}, {7})).Display()->AsString();
+   // std::string dfRDSZAsString;
+   // ROOT_EXPECT_ERROR(dfRDSZAsString = dfRDSZ->AsString(), "TTreeReader::SetEntriesRange()",
+   //                   "first entry out of range 0..5");
+   EXPECT_EQ(dfRDSZ, dfSimple);
+
+   EXPECT_THROW(
+      try {
+         RDatasetSpec("treeA", {"file0.root"s, "file1.root"s}, {9, 7});
+      } catch (const std::logic_error &err) {
+         EXPECT_EQ(std::string(err.what()), "RDatasetSpec: fStartEntry cannot be larger than fEndEntry.");
+         throw;
+      },
+      std::logic_error);
+   EXPECT_THROW(
+      try {
+         RDatasetSpec("treeA", {"file0.root"s, "file1.root"s}, {9, 2});
+      } catch (const std::logic_error &err) {
+         EXPECT_EQ(std::string(err.what()), "RDatasetSpec: fStartEntry cannot be larger than fEndEntry.");
+         throw;
+      },
+      std::logic_error);
+
    gSystem->Exec("rm file0.root file1.root file2.root");
 }
+
+TEST(RDatasetSpecTest, Friends)
```

Comment:
Nice that you added these tests with friend trees, but I think it's a bit too convoluted for a single test. Maybe you can have a smaller example before this one and add a table representation of the dataset as a comment in this long one